### ISSUE TITLE
[feat] Windows - Build MSIX Packages to Support Package Extensions

### ISSUE THREAD

**Issue by madaster97** [tauri-apps/tauri#4818]
### Describe the problem

There are a number of windows features that require a desktop app to have [package identity](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/modernize-packaged-apps). For example, you can use a [package extension](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-extensions) to have websites redirect to your desktop application ([app URI handlers](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/web-to-app-linking)) similar to universal links on iOS.

The issue is that Tauri apps build to msi which cannot have package identity, so we can't use the features above in our apps. We'd need to build to MSIX to do this.

## Enabling Package Extensions from Config
If we implemented package identity under the hood (as in, build msix packages), then it'd be great to give config options for some useful [package extensions](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-extensions), such as [`windows.appUriHandler`](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/web-to-app-linking). This would for instance increase OAuth 2.0 support in Tauri, since we'd have a method other than [custom URI schemes](https://github.com/tauri-apps/tauri/issues/323) or localhost servers to accept OAuth 2.0 requests from the system default browser. [Discord thread](https://discordapp.com/channels/616186924390023171/990632073322721341/990723231440855121) for reference.

## Option for Package Identity: Sparse Package
Microsoft outlines 2 ways to obtain package identity:

- Build MSIX directly from your [source code](https://docs.microsoft.com/en-us/windows/msix/desktop/source-code-overview) (discussed as an alternative below)
- Build a [sparse package](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps), which notably is ⚠ only supported in Windows 10 version 2004 and later

A sparse package effectively sits beside your existing apps and adds additional features using an MSIX that points at your existing MSI based app. It seems like a much lower lift than building MSIX from source, for some reasons discussed in the alternatives section below.

Microsoft has a nice [tutorial](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps) outlining how to generate a sparse package, and here's where I expect the complexity to be for Tauri:
- The [additional build step](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps#build-and-sign-the-sparse-package) includes signing the package. We'd need a system for choosing/generating a certificate to sign the package with
- There is a [run-time registration step](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps#register-your-sparse-package-at-run-time) we'd need to introduce. I'm not actually sure if this goes in the MSI or MSIX 🤷‍♂️. That tutorial includes a code snippet, and WiX has some [documentation](https://wixtoolset.org/documentation/manual/v3/wixdev/extensions/authoring_custom_actions.html) on injecting the action with a template. We'd just need to write a dll for this and include it in the %AppData%\Local\tauri\WixTools folder (I think)

### Alternatives considered

Creating [full MSIX packages](https://docs.microsoft.com/en-us/windows/msix/desktop/source-code-overview) instead of or in addition to MSI. This is nice because it's supported before ✔Windows 10 version 2004, but has a couple big drawbacks ❌:

- Wix doesn't currently support [MSIX](https://github.com/wixtoolset/issues/issues/5781), so we'd have to re-tool to one of the techniques in the link above
- There are a ton of [prereqs](https://docs.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-prepare) Tauri would need to meet in terms of what windows features it uses (example: Using a [background task](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/create-and-register-a-background-task) instead of a windows service)
- [ ] I would appreciate if someone could review the prereqs list

### Additional context

I researched this because it's important for flexibility in using Tauri as an OAuth 2.0 client app. In [issue 323](https://github.com/tauri-apps/tauri/issues/323) I saw some confusion on this and wanted to call out based on [RFC 8252 section 8.12](https://datatracker.ietf.org/doc/html/rfc8252#section-8.12) you shouldn't be conducting OAuth 2.0 from a web-view. Instead, you should be launching to the system default browser and having it redirect back to the initial native app. Adding app URI Handler support would be a great way for Tauri apps to conduct this redirect.

**Comment by lucasfernog** [comment]
Maybe we can do that as part of the NSIS implementation.

**Comment by amrbashir** [comment]
I have plans to make a MSIX bundle for tauri but probably in v3 (unless someone else wants to do it before that).

If you just want to have deep-link, then we can add it for both WiX and NSIS once #4674 is merged.

**Comment by madaster97** [comment]
Thanks both!

@amrbashir - Yeah getting a custom URI scheme with WiX or NSIS would be great. I see that sticking around for some time, precisely because people won't have to deal with MSIX files.

**Comment by amrbashir** [comment]
I leave this issue open for msix but there is already an issue for deep linking here #323

**Comment by gezihuzi** [comment]
When I submitted the `msi` application generated by Tauri to the Microsoft Store, I received the following message:

If your EXE or MSI cannot comply with [Microsoft Store policy 10.2.9](https://go.microsoft.com/fwlink/?linkid=2262002), you can consider repackaging your existing EXE or MSI to MSIX format. Microsoft Store offers many complimentary benefits for MSIX format such as code signing, hosting etc. For more details, [click here](https://go.microsoft.com/fwlink/?linkid=2261818).

You can run your EXE or MSI installers through [MSIX Packaging Tool ](https://go.microsoft.com/fwlink/?linkid=2261788)and obtain an MSIX package that you can submit to the Microsoft Store as [MSIX packaged app](https://go.microsoft.com/fwlink/?linkid=2247663). Note that you have to [delete your app name](https://go.microsoft.com/fwlink/?linkid=2189821) from existing Win32 app in Partner Center in case you want to use the same for MSIX packaged app.

---

I manually converted the `msi` package to an `msix` installation package using the `MSIX Packaging Tool`. I will try to submit this content to the Microsoft Store and provide updates on progress.

**Comment by eythaann** [comment]
@gezihuzi Hi, I want to do the same, do you need to sign the msix or can I upload it to MS Store without sign it? I read that Microsoft sign the msix installers on the store but to upload it for review it stills need to be signed.

**Comment by gezihuzi** [comment]
> @gezihuzi Hi, I want to do the same, do you need to sign the msix or can I upload it to MS Store without sign it? I read that Microsoft sign the msix installers on the store but to upload it for review it stills need to be signed.

To my knowledge, there are currently two options for creating products in the Microsoft Store:

1. `MSIX and PWA`, MSIX supports packages uploaded without requiring signing, as Microsoft Store automatically signs them.

2. `EXE and MSI`, which require purchasing a certificate for signing.

If you are only publishing to the Microsoft Store, choose the first option and submit an unsigned package created using the `MSIX Packaging Tool`.

---

The package I submitted has been uploaded and approved. I will write an article later on how I submitted a `Tauri` app to the Microsoft Store.

**Comment by eythaann** [comment]
Cool I did the same, now I'm waiting for certification or approval. And yes an article sounds great currently there are no articles or something talking about msix with tauri.

**Comment by gezihuzi** [comment]
@eythaann Hi, you can check this, https://gezihuzi.github.io/2024/04/05/publish-tauri-app-to-ms-store.html

**Comment by GhostBone** [comment]
> @eythaann Hi, you can check this, https://gezihuzi.github.io/2024/04/05/publish-tauri-app-to-ms-store.html

非常感谢！

**Comment by joeyoropesa-dev** [comment]
What about appx? Compiling into appx UWP app?

**Comment by wayneforrest** [comment]
A vote comment - for this feature.

> Please add your vote for this feature with a thumbs up.

**Comment by petrvecera** [comment]
Hi guys, this is really needed feature. It's the reason why we had to go back to Electron with some of our other apps as there the .msix installers and Microsoft Store publishing is much more easier.

**Comment by Choochmeque** [comment]
Hi everyone,

I've pushed a small tool that I'm using internally to publish my Tauri app to the Microsoft Store - hopefully it'll be useful for others too:

https://github.com/Choochmeque/msixbundle-rs

But of course, Tauri should ideally support this out of the box - especially generating the AppxManifest.xml automatically based on information from tauri.conf.json.

Technically, it shouldn’t be too complicated to implement.

**Comment by soywiz** [comment]
@Choochmeque Thanks for the tool!

Could you explain how to use it for creating a x64 msix from a tauri project?

**Comment by Choochmeque** [comment]
> [@Choochmeque](https://github.com/Choochmeque) Thanks for the tool!
> 
> Could you explain how to use it for creating a x64 msix from a tauri project?

I wrote a better tool for tauri (I am using it internally): https://github.com/Choochmeque/tauri-windows-bundle

"One" command to build msix bundle. App manifest and multiarch support included. Check the instructions, but it should be pretty straightforward.

Cheers!

**Comment by soywiz** [comment]
@Choochmeque Awesome. Thank you very much ^_^!

**Comment by jaime-rivas** [comment]
Hi everyone! I’m reviewing our Windows distribution pipeline and wanted to check in with the maintainers about the current stance on MSIX support. From the discussion above, I understand it was previously considered out of scope. I also saw a mention that it might be reconsidered for v3. I’m just trying to understand whether that’s still the case so I can plan accordingly.

No expectations, pressure at all, just looking for clarity. Thank you for your time and for all the work you put into Tauri.

**Comment by FabianLars** [comment]
No explicit plans, we do want it, but it's not the highest priority. We also don't have an ETA for v3 itself so even if that were the target release, it wouldn't have a date.

**Comment by yunmoxinghe** [comment]
Looking forward to native support for MSIX soon.

**Comment by yunmoxinghe** [comment]
Looking forward to an MSIX build with native support and automatic updates.
